Multithreading 02

td-extreme opened the pull request.

td-extreme committed:
Merge pull request #20 from td-extreme/refactor_for_multithread

Refactor for multithread

td-extreme committed:
Added multithreading.

Moved Running Loop into HttpServerRunner

Created HttpServerThread.
  Moved logic from HttpServerRunner into HttpServerThread

td-extreme committed:
test

td-extreme committed:
Multithreading with integration test

td-extreme committed:
Clean up Multithread test

Broke the ClientSocketIO class up into three classes.

Created Mocks for each of the separate interfaces for ClientSocket

Made the names of the tests threads clearer

Removed Mocking of all the classes in test except the ClientSocket

td-extreme committed:
Added Mock used for MultiThreadTest

td-extreme committed:
Removed dead file

td-extreme committed:
Moved ThreadTest from integration to test

td-extreme committed:
Added BlackBox test for multithreading.

dadarek commented on src/blackboxTest/bash/BlackBoxTest.sh:
```diff
@@ -27,39 +28,18 @@ dispayResults() {
-  rm ${TESTPATH}/$2
-  rm ${TESTPATH}/$3
+updateExitCode() {
+if [ $? -eq 1 ]; then
```
Not all programs return a `1` on failure. Typically the convention is 0 on success and non-0 on failure. Maybe a shell equivalent of `result ||= $?` would be better here–would get rid of the if check.

td-extreme replied:
Could I change this instead to if [ $? `not equal to` 0 ]; then

dadarek replied:
probably

dadarek commented on src/blackboxTest/bash/BlackBoxTest.sh:
```diff
@@ -27,39 +28,18 @@ dispayResults() {
 setUp
-testGetDirectoryContents
-testGetFileContents
+./TestGetRequest.sh ${TESTPATH}
```
Nice splitting this up

dadarek commented on src/blackboxTest/bash/TestGetRequest.sh:
```diff
@@ -0,0 +1,42 @@
+
+
+
+testResults() {
```
This is actually more like `printTestResults` or `printSummary`, or something like that, right? The name currently suggests that it's actually running something.

dadarek commented on src/blackboxTest/bash/TestMultiThreading.sh:
```diff
@@ -0,0 +1,45 @@
+
+
+runThreads() {
+  ./requestWithDelay.sh ${TESTPATH} > /dev/null 2>/dev/null &
```
Could _probably_ cleanup a little duplication here. The two files are almost identical except that they call two different files that are also almost identical. The only thing that differs there is the output file to which they write and the duration which they sleep.

Why the IO redirect?

td-extreme replied:
Right, should probably just pass those in as parameters.

I did the redirect because it didn't seem necessary to have all the text form the telnet commands and then the returned response print to the screen. The person running the test doesn't need to see that.

dadarek replied:
Got it.

dadarek commented on src/blackboxTest/bash/TestMultiThreading.sh:
```diff
@@ -0,0 +1,45 @@
+# Main
+
+runThreads
+sleep 12
```
Although sleeps are sometimes the only way out, look for and prefer other mechanisms for synchronization when possible.

In this case, what you're really looking for is the existence (and content) of those two files. Should be pretty simple with something like `wc -l` or something of that sort.

Don't worry about changing it here.

dadarek commented on src/main/java/HttpServerThread.java:
```diff
@@ -0,0 +1,48 @@
+package com.td.HttpServer;
+
+public class HttpServerThread implements Runnable {
```
This works, _but_, your entire `HttpServerThread` class is now tied to the java threading library. That's typically fine (actually you _have_ to implement `Runnable` to use threads) but try to look for better ways to corner this dependency.

`HttpServerThread` already has one responsibility (coordinating the retrieval of a request, fetching a corresponding handler for it, and then piping the response back), and mixing this in does not feel right.

Is it possible to separate all of this "coordination" code from the actual "running this thing somewhere (in a thread)"?

dadarek replied:
If it helps, think about what would you have to do if I asked you for an easy way to configure this server to be multithreaded or not?

Ideally, you would change one line of code to enable/disable that feature (hint: instantiate _this_ guy for single threading or _that_ guy for multithreading).

dadarek commented on src/main/java/HttpServerRunner.java:
```diff
@@ -1,44 +1,33 @@
+        continue;
+      }
+      HttpServerThread httpServerThread = new HttpServerThread(client, handlerRouter, httpRequestBuilder, httpResponseWriter);
+      Thread thread = new Thread(httpServerThread);
```
Similar idea here. `Thread` is an API class. `ClientSocketIO`, `HttpServerThread`, `ServerSocket`, and other several classes in here are all your code (creamy center).

Can you rip this API dependency out of here?

Taking it further - can you structure your hierarchy in such a way that "nobody knows" (and nobody cares) that this thing is actually multithreaded, _except_ for the thing that's actually responsible for the multithreading? Can you isolate the "mutlithreading" responsibility into one class?

td-extreme committed:
Seperated multithreading out of connection hander

td-extreme committed:
made test Verbose for debug

td-extreme committed:
Fixed test for exit code

dadarek commented on src/test/java/MockClientSocket.java:
```diff
@@ -0,0 +1,42 @@
+
+  public void closeClientConnection() {
+    sleep();
+    System.out.println("Closing connectoin at : " + System.currentTimeMillis());
```
probably don't need this

dadarek replied:
probably don't need this

td-extreme committed:
Chagned test to use netcat

dadarek commented on src/test/java/MultiThreadTest.java:
```diff
@@ -0,0 +1,38 @@
+    connectionProcessRunnerMutliThread.execute(connectionLongDelay);
+    connectionProcessRunnerMutliThread.execute(connectionNoDealy);
+    Thread.sleep(1500);
+    System.out.println("no delay end time  : " + clientNoDelay.getClosedTime());
```
Same here - not sure if this is debug or intentionally left in as information?

dadarek replied:
Same here - not sure if this is debug or intentionally left in as information?

dadarek commented:
Assuming CI passes, I'm 👍 on this

td-extreme committed:
Removed debug output

td-extreme committed:
Updated function so exit code is passed in.

The function needs to make an if check and then still use the exit code
value from the last test. It is now passed in instead so that after the
if check, the function can use the exit code from the last test.

mikeknep commented:
+1
